### PR TITLE
Fix immutability-helper

### PR DESCRIPTION
**Pull request by gordon-to**
It no longer defaults to any[] as the return value of update

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [ ] The package does not provide its own types, and you can not add them.
- [ ] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, and `strictNullChecks` set to `true`.

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [x] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

If removing a declaration:
- [ ] If a package was never on DefinitelyTyped, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

File: types/immutability-helper/index.d.ts
Comment (andy-ms) [inline types/immutability-helper/index.d.ts]: `K` not used. You might add a `tslint.json` containing `{ "extends": "dtslint/dt.json" }` to catch this kind of error.

File: types/immutability-helper/index.d.ts
Comment (andy-ms) [inline types/immutability-helper/index.d.ts]: Add your name to the end of the list: `// Definitions by: Sean Kelley <https://github.com/seansfkelley>, Adam Gordon <https://github.com/xmrwhite>`

File: types/immutability-helper/immutability-helper-tests.ts
Comment (seansfkelley) [inline types/immutability-helper/immutability-helper-tests.ts]: Might as well leave this, it's handy documentation to whoever comes to read the tests. Would be nice if we could have proper negative tests though (https://github.com/DefinitelyTyped/DefinitelyTyped/issues/6987).

File: types/immutability-helper/immutability-helper-tests.ts
Comment (seansfkelley) [inline types/immutability-helper/immutability-helper-tests.ts]: Can you make this test file consistently formatted? Either commented blocks or namespaced blocks, but mixing both is noisy.

File: types/immutability-helper/index.d.ts
Comment (seansfkelley) [reply types/immutability-helper/index.d.ts]: Given that the test code above pulled from the React docs has a heterogeneous array, that's probably something we want to support? Perhaps `<K extends Array<any>>(value: K, ...): K`?

File: types/immutability-helper/immutability-helper-tests.ts
Comment (seansfkelley) [inline types/immutability-helper/immutability-helper-tests.ts]: I think these tests are probably still useful and should be kept around.

**Review by seansfkelley — CHANGES_REQUESTED** [review]

File: types/immutability-helper/index.d.ts
Comment (xMrWhite) [reply types/immutability-helper/index.d.ts]: `<K extents Array<any>>` will cause type erasure. There is no way to specify a generic as `<_ extends Array<T>`.  So either K or T will be unused unless i cast value as `K` and return as `T[]`

File: types/immutability-helper/immutability-helper-tests.ts
Comment (xMrWhite) [reply types/immutability-helper/immutability-helper-tests.ts]: They don't test anything type related. immutability-helper has it's own test.

File: types/immutability-helper/index.d.ts
Comment (xMrWhite) [reply types/immutability-helper/index.d.ts]: looks like `<_ extends Array<T>` makes intellij happy, unfortunately the project won't build locally at all so i'll wait for another travis run.
If you pass in an `Array<any>` or a heterogeneous array you will get out the same, but it won't erase the type of a typed array.

Comment (seansfkelley) [reply types/immutability-helper/index.d.ts]: I don't know what you mean by "cause type erasure". Can you show me examples where the types don't behave as expected if you do what Andy or I suggest? (Also, I should have named the type variable `T` in my example, not `K`, since there is no other type variable).

File: types/immutability-helper/immutability-helper-tests.ts
Comment (seansfkelley) [reply types/immutability-helper/immutability-helper-tests.ts]: This _is_ the immutability-helper test suite, so I'm not sure what you mean by "[it] has its own test". It's a drop-in replacement for `react-addons-update`, so it seems reasonable to include the basic docs examples as a smoke test in addition to any more-thorough type-heavy tests. I don't think these tests are extraneous.

File: types/immutability-helper/index.d.ts
Comment (andy-ms) [reply types/immutability-helper/index.d.ts]: To clarify, my suggestion would be to simply remove `K`.

**Comment by RyanCavanaugh** [comment]
@xMrWhite Please fix the failures indicated in the Travis CI log.

**Comment by xMrWhite** [comment]
@RyanCavanaugh Error message is not super helpful. Though I fixed it based on @andy-ms 's comment.
```javascript
// Definitions by: Sean Kelley <https://github.com/seansfkelley>
// Definitions by: Adam Gordon <https://github.com/xmrwhite>
```
to 
```javascript
// Definitions by: Sean Kelley <https://github.com/seansfkelley>, Adam Gordon <https://github.com/xmrwhite>
```
```bash
> definitely-typed@0.0.1 test /Users/ag/projects/DefinitelyTyped
> node node_modules/types-publisher/bin/tester/test.js --run-from-definitely-typed

Clean data
Clean logs
Clean output
# Typing Publish Report Summary
Started at Tue, 31 Oct 2017 21:36:47 GMT
Found 3870 typings folders in /Users/ag/projects/DefinitelyTyped/types
Error: At 3:65 : Expected one of
	/(, )|(,?\r?\n\/\/\s\s+)/
	/\r?\n\/\/ Definitions: [^\r\n]+/
    at Object.parseHeaderOrFail (/Users/ag/projects/DefinitelyTyped/node_modules/definitelytyped-header-parser/index.js:45:15)
    at /Users/ag/projects/DefinitelyTyped/node_modules/types-publisher/src/lib/definition-parser.ts:93:3
    at Generator.next (<anonymous>)
    at fulfilled (/Users/ag/projects/DefinitelyTyped/node_modules/types-publisher/bin/lib/definition-parser.js:4:58)
    at <anonymous>
npm ERR! Test failed.  See above for more details.
npm test  26.37s user 2.97s system 110% cpu 26.538 total
```

**Comment by RyanCavanaugh** [comment]
@seansfkelley - Thanks for your review of this PR! Can you please look at the new code and update your review status if appropriate?

**Comment by seansfkelley** [comment]
@xMrWhite Can you address the comments I've left?

**Comment by RyanCavanaugh** [comment]
This PR has been open and unchanged 5 days without signoff or complaint. This will be merged by a maintainer soon if there are no objections.

**Review by seansfkelley — CHANGES_REQUESTED** [review]
As noted in previous review.

**Comment by seansfkelley** [comment]
@RyanCavanaugh I marked this with "changes requested" when it was first opened, and that is still accurate, as my comments have not been responded to. I just did it again to ensure I had not erroneously marked at as just "comments". I think your bot (?) has a bug.

@xMrWhite Please address the comments I've left.

**Comment by minestarks** [comment]
I'm closing this PR due to staleness, as the revision requests haven't been addressed in almost two months. If you'd still like to get these changes in, please open a new PR. Thanks!

**Pull request closed without merging**
